api/MarbleCollectorApi/Controllers/GrantsController.cs

```diff
@@ -1,59 +1,137 @@
-﻿using Microsoft.AspNetCore.Http;
+﻿using System;
+using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Mvc;
 using System.Collections.Generic;
+using System.Linq;
+using MarbleCollectorApi.Data.Mapping;
+using MarbleCollectorApi.Data.Repository;
 using MarbleCollectorApi.ViewModels;
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Microsoft.AspNetCore.SignalR;
+using MarbleCollectorApi.Hubs;
+using Microsoft.AspNetCore.Authorization;
 
 namespace MarbleCollectorApi.Controllers
 {
+    [Authorize]
     [Route("api/[controller]")]
     [ApiController]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
     public class GrantsController : Controller
     {
+        private readonly IGrantRepository _grantRepository;
+        private readonly IHubContext<ParentNotificationHub> _parentNotificationHubContext;
+        private readonly IHubContext<ChildrenNotificationHub> _childrenNotificationHubContext;
+
+        public GrantsController(IGrantRepository grantRepository, IHubContext<ParentNotificationHub> parentNotificationHubContext, IHubContext<ChildrenNotificationHub> childrenNotificationHubContext)
+        {
+            _grantRepository = grantRepository ?? throw new ArgumentNullException(nameof(grantRepository));
+            _parentNotificationHubContext = parentNotificationHubContext;
+            _childrenNotificationHubContext = childrenNotificationHubContext;
+        }
+
         // TODO js (04.03.2021): Can all users get all grants?
         [HttpGet]
         [ProducesResponseType(StatusCodes.Status200OK)]
         public ActionResult<IEnumerable<Grant>> GetGrants()
         {
-            return NotFound(); // TODO js (04.03.2021): To be implemented!
+            return Ok(_grantRepository.GetAll().Select(grant => grant.Map()));
         }
 
         [HttpGet("{id}")]
         [ProducesResponseType(StatusCodes.Status200OK)]
         public ActionResult<Grant> GetGrantById(int id)
         {
-            return NotFound(); // TODO js (04.03.2021): To be implemented!
+            var grant = _grantRepository.GetSingle(id);
+            if (grant == null)
+            {
+                return NotFound();
+            }
+
+            return Ok(grant.Map());
         }
 
         [HttpPost()]
+        [Authorize(Roles = Const.UserRoleParent)]
         [ProducesResponseType(StatusCodes.Status201Created)]
         [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
         public ActionResult<Grant> CreateGrant(Grant grant)
         {
-            return NotFound(); // TODO js (04.03.2021): To be implemented!
+            var entityEntry = _grantRepository.Add(grant.Map());
+
+            try
+            {
+                _grantRepository.Commit();
+            }
+            catch
+            {
+                // TODO hs 210307, is i.e. foreign key constraint a bad request or internal server error?
```

Comment:
@StuderS41 Das könnte auch einen 403 zurückliefern (aus Wikipedia):
403 Forbidden
The request contained valid data and was understood by the server, but the server is refusing action. This may be due to the user not having the necessary permissions for a resource or needing an account of some sort, or attempting a prohibited action (e.g. creating a duplicate record where only one is allowed). This code is also typically used if the request provided authentication by answering the WWW-Authenticate header field challenge, but the server did not accept that authentication. The request should not be repeated.